Unsafe Extern Blocks (#3484)
rust-lang/rfcs: text/0000-unsafe-extern-blocks.md

Within the block you can declare the exernal functions and statics that you want to make visible within the current scope.
Each function declaration gives only the function's signature, similar to how methods for traits are declared.
If calling a foreign function is `unsafe` then you must declare the function as `unsafe fn`, otherwise you can declare it as a normal `fn`.
Each static declaration gives the name and type, but no initial value.

* If the `unsafe_code` lint is denied or forbidden at a particular scope it will cause the `unsafe extern` block to be a compilation error within that scope.
* Declaring an incorrect external item signature can cause Undefined Behavior during compilation, even if Rust never accesses the item.

```rust
unsafe extern {
    // sqrt (from libm) can be called with any `f64`
    pub fn sqrt(x: f64) -> f64;
    
    // strlen (from libc) requires a valid pointer,
    // so we mark it as being an unsafe fn
    pub unsafe fn strlen(p: *const c_char) -> usize;
    
    pub static IMPORTANT_BYTES: [u8; 256];
    

>     pub static LINES: UnsafeCell<i32>;

Reviewer: ```suggestion
    pub static LINES: SyncUnsafeCell<i32>;
```

`UnsafeCell` isn't `Sync`

Rust can utilize functions and statics from foreign code that are provided during linking, though it is `unsafe` to do so.

An `extern` block can be placed anywhere a function declaration could appear (generally at the top level of a module).
* You can always write `unsafe extern { ... }`.
* If the `unsafe_code` lint is denied or forbidden at a particular scope it will cause the `unsafe extern` block to be a compilation error within that scope.
* On editions >= 2024, you must write all `extern` blocks as `unsafe extern`.
* On editions < 2024, it is allowed to write an `extern` block *without* the `unsafe` keyword, but this generates a compatibility warning that you should use the `unsafe` keyword.

Within an `extern` block is zero or more declarations of external functions and/or external static values.
An extern function is declared with a `;` instead of a function body (similar to a method of a trait).
An extern static value is also declared with a `;` instead of an expression (similar to an associated const of a trait).
In both cases, the actual function body or value is provided by whatever external source (which is probably not even written in Rust).

> When an `extern` block is used (with or without `unsafe` in front of it), all declarations within that `extern` block should have the `unsafe` or `safe` keywords as part of their signature.

Reviewer 2: @Lokathor updated the text, and this looks correct now.

If one of the two keywords is not explicitly provided, the declaration is assumed to be `unsafe`.
The `safe` keyword is a contextual keyword, only used within `extern` blocks.

```rust
unsafe extern {
    // sqrt (from libm) can be called with any `f64`
    pub safe fn sqrt(x: f64) -> f64;
    
    // strlen (from libc) requires a valid pointer,
    // so we mark it as being an unsafe fn
    pub unsafe fn strlen(p: *const c_char) -> usize;

    // this function doesn't say safe or unsafe, so it defaults to unsafe
    pub fn free(p: *mut core::ffi::c_void);
    
    pub safe static IMPORTANT_BYTES: [u8; 256];
    
    pub safe static LINES: SyncUnsafeCell<i32>;
}
```

`extern` blocks are `unsafe` because if the declaration doesn't match the actual external function, or the actual external data, then it causes compile time Undefined Behavior (UB).

Once they are unsafely declared, a `safe` item can be used outside the `extern` block as if it were any other safe function or static value declared within rust.

> The unsafe obligation of ensuring that the correct items are being linked to is performed by the crate making the declaration, not the crate using of that declaration.

Reviewer: ```suggestion
The unsafe obligation of ensuring that the correct items are being linked to is performed by the crate making the declaration, not the crate using that declaration.
```

- Feature Name: `unsafe_extern`
- Start Date: 2023-05-23
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

In Edition 2024 it is `unsafe` to declare an `extern` function or static, but external functions and statics *can* be safe to use after the initial declaration.

# Motivation
[motivation]: #motivation

Simply declaring extern items, even without ever using them, can cause Undefined Behavior.

> When performing cross-language compilation, attributes on one function declaration can flow to the foreign declaration elsewhere within LLVM and cause a miscompilation.

Reviewer 3: I think I somewhat understand why this is the case in LLVM, as they have no concept of `unsafe`, and in the end the function will have only one signature, so might as well merge them and their attributes.

But maybe the RFC text could expand a little more on why this is not just a bug to fix in LLVM, or in `rustc`'s usage of LLVM, perhaps with links to open issues? Maybe an alternative fix would be to never expose `extern "C" { fn foo(); }` to LLVM if it's unused?

Reviewer 4: In fact I think the state of whether and where exactly this is UB or whether there's a bug is currently open -- [here](https://github.com/rust-lang/rust/issues/46188#issuecomment-1829256377) and [here](https://github.com/rust-lang/unsafe-code-guidelines/issues/198#issuecomment-1857384321).

Reviewer 2: I've incorporated a link to that first issue into the document (the second issue is linked from the first).

In Rust we consider all sources of Undefined Behavior to be `unsafe`, and so we must make declaring extern blocks be `unsafe`.
The up-side to this change is that in the new style it will be possible to declare an extern fn that's safe to call after the initial unsafe declaration.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Rust can utilize functions and statics from foreign code that are provided during linking, though it is `unsafe` to do so.

An `extern` block can be placed anywhere a function declaration could appear (generally at the top level of a module).

* On editions >= 2024, you *must* write all `extern` blocks as `unsafe extern`.

> * On editions < 2024, you *may* write `unsafe extern`, or you can write an `extern` block without the `unsafe` keyword. Writing an `extern` block without the `unsafe` keyword is provided for compatibility only, and will generate a warning.

Reviewer 5: This should not immediately start producing a warning in older editions; that would be a large change. If we do this at all, we should phase it in.

Author: The rate at which this is rolled out isn't really an issue, but I believe that we should get there at some point.

Reviewer 2: In the RFC, I've added the word "eventually" to better capture what we all probably understood here.

* `unsafe extern` interacts with the `unsafe_code` lint, and a `deny` or `forbid` with that lint will deny or forbid the unsafe external block.

Within an `extern` block is zero or more declarations of external functions and/or external static values.
An extern function is declared with a `;` instead of a function body (similar to a method of a trait).
An extern static value is also declared with a `;` instead of an expression (similar to an associated const of a trait).
In both cases, the actual function body or value is provided by whatever external source (which is probably not even written in Rust).

When an `unsafe extern` block is used, all declarations within that `extern` block *must* have the `unsafe` or `safe` keywords as part of their signature.

> The `safe` keyword is a contextual keyword, it is currently only used within `extern` blocks.

Reviewer 6: ```suggestion
The `safe` keyword is a contextual keyword, it is currently allowed used within `extern` blocks.
```

Reviewer 2: Fixed.

If an `extern` block is used in an older edition without the `unsafe` keyword, declarations *cannot* specify `safe` or `unsafe`.
Code must update to `unsafe extern` style blocks if it wants to make `safe` declarations.

```rust
unsafe extern {
    // sqrt (from libm) can be called with any `f64`
    pub safe fn sqrt(x: f64) -> f64;
    
    // strlen (from libc) requires a valid pointer,
    // so we mark it as being an unsafe fn
    pub unsafe fn strlen(p: *const c_char) -> usize;

>     // this function doesn't say safe or unsafe, so it defaults to unsafe
>     pub fn free(p: *mut core::ffi::c_void);

Reviewer 6: ```suggestion
    // this function doesn't say safe or unsafe, so it causes a compilation error.
    // on an older edition in an `extern{}` block it would be assumed to be unsafe
    pub fn free(p: *mut core::ffi::c_void); //~ error: a function in an extern block must be marked as either `unsafe` or `safe`
```

Reviewer 2: The proposed change here is not correct.  According to the rules that pnkfelix summarized:

> - An item without a `safe` nor `unsafe` marker within an `extern` block is implicitly unsafe (i.e. this is the current Rust 2015 semantics).

    
    pub safe static IMPORTANT_BYTES: [u8; 256];
    
    pub safe static LINES: SyncUnsafeCell<i32>;
}
```

> `extern` blocks are `unsafe` because if the declaration doesn't match the actual external function, or the actual external data, then it causes compile time Undefined Behavior (UB).

Reviewer 6: What "compile time UB" mean here?

Reviewer 2: I've clarified this wording.

Once they are unsafely declared, a `safe` item can be used outside the `extern` block as if it were any other safe function or static value declared within rust.
The unsafe obligation of ensuring that the correct items are being linked to is performed by the crate making the declaration, not the crate using that declaration.

Items declared as `unsafe` *must* still have a correctly matching signature at compile time, but they *also* have some sort of additional obligation for correct usage at runtime.
They can only be used within an `unsafe` block.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> The grammar of the langauge is updated so that:

Reviewer 7: ```suggestion
The grammar of the language is updated so that:
```

Reviewer 2: Fixed, thanks for this.

* Editions >= 2024 *must* prefix all `extern` blocks with `unsafe`.

> * Editions < 2024 *should* prefix `extern` blocks with `unsafe`, this is a warn-by-default compatibility lint when `unsafe` is missing.

Reviewer 6: Personally I'm fine with this, but it's also worth noting that this will warn on **all `extern` blocks everywhere**, which is a bit destructive and will probably make users angry (no one likes CI being broken by lints).

Making this allow-by-default at the start might be a good idea. (in _addition_ to making sure it has machine applicable solution).

Reviewer 2: In the RFC, I've added the word "eventually" to better capture what we all probably understood here.

> * Editions < 2024 *should* prefix `extern` blocks with `unsafe`, this is a warn-by-default compatibility lint when `unsafe` is missing.

Reviewer 8: the intention was to make function declarations consistent with function definitions ("unsafe" are explicitly annotated), and "eventually" all unannotated extern function declaration would be "safe", so the extern block itself has to be marked as "unsafe" to specify that safety has been checked

i'm not convinced this is a valid reason to force `unsafe` on *every* `extern` block. at the very least, if the block contains *only* `unsafe` items, there is no reason mark to `extern` block itself again, it is like `unsafe { unsafe { expr; } }` is an `unused_unsafe` warning.

i'm also not convinced that "eventually" killing the `safe` context keyword is a proper path to take, since an FFI fn declaration is a blackbox allowed to do any nasty thing, unlike a fn definition which `rustc` can inspect into the implementation to mechanically check its safety.

> * Editions < 2024 *should* prefix `extern` blocks with `unsafe`, this is a warn-by-default compatibility lint when `unsafe` is missing.

Reviewer 2: We've added the word "eventually" to this section as a clarification about this warning.  The lang team is conscious of not creating too much noise from new warnings and will act carefully here during stabilization of any changes.

In terms of the rationale for these changes, see in particular [this comment](https://github.com/rust-lang/rfcs/pull/3484#issuecomment-1758275493) from **pnkfelix**.

> Replace the *Functions* and *Statics* sections with the following:

Reviewer 6: Sections of what does this replace?

Reviewer 2: Thanks, I've clarified this.

### Functions

> Functions within external blocks are declared in the same way as other Rust functions, with the exception that they must not have a body and are instead terminated by a semicolon. Patterns are not allowed in parameters, only IDENTIFIER or _ may be used. The function qualifiers `const`, `async`, and `extern` are not allowed. If the function is unsafe to call, then the function should use the `unsafe` qualifier. If the function is safe to call, then the function should use the `safe` qualifier (a contextual keyword). Functions that are not qualified as `unsafe` or `safe` are assumed to be `unsafe`.

Reviewer 6: ```suggestion
Functions within external blocks are declared in the same way as other Rust functions, with the exception that they must not have a body and are instead terminated by a semicolon. Patterns are not allowed in parameters, only IDENTIFIER or _ may be used. The function qualifiers `const`, `async`, and `extern` are not allowed. If the function is unsafe to call, then the function must use the `unsafe` qualifier. If the function is safe to call, then the function should use the `safe` qualifier (a contextual keyword). Functions must be qualified as either `unsafe` or `safe`, a missing qualifier is a compilation error.
```
Not _exactly_ sure about the wording, but this seems to be correct wrt recent changes to the RFC.

Reviewer 2: The suggested change is not correct.  According to the RFC and the lang consensus, omitting the `safe` or `unsafe` qualifier is allowed.

If the function signature declared in Rust is incompatible with the function signature as declared in the foreign code it is Undefined Behavior to compile and link the code.

Functions within external blocks may be called by Rust code, just like functions defined in Rust. The Rust compiler will automatically use the correct foreign ABI when making the call.

When coerced to a function pointer, a function declared in an extern block has type
```rust
extern "abi" for<'l1, ..., 'lm> fn(A1, ..., An) -> R
```
where `'l1`, ... `'lm` are its lifetime parameters, `A1`, ..., `An` are the declared types of its parameters and `R` is the declared return type.

### Statics

> Statics within external blocks are declared in the same way as statics outside of external blocks, except that they do not have an expression initializing their value. If the static is unsafe to access, then the static should use the `unsafe` qualifier. If the static is safe to access (and immutable), then the static should use the `safe` qualifier (a contextual keyword). Statics that are not qualified as `unsafe` or `safe` are assumed to be `unsafe`.

Reviewer 6: ```suggestion
Statics within external blocks are declared in the same way as statics outside of external blocks, except that they do not have an expression initializing their value. If the static is unsafe to access, then the static must use the `unsafe` qualifier. If the static is safe to access (and is immutable), then the static should use the `safe` qualifier (a contextual keyword). Statics must qualified as either `unsafe` or `safe`, a missing qualifier is a compilation error.
```

Reviewer 2: The suggested change is not correct.  According to the RFC and the lang consensus, omitting the `safe` or `unsafe` qualifier is allowed.

> Extern statics can be either immutable or mutable just like statics outside of external blocks. An immutable static must be initialized before any Rust code is executed. It is not enough for the static to be initialized before Rust code reads from it. A mutable extern static is always `unsafe` to access, the same as a Rust mutable static.

Reviewer 6: ```suggestion
Extern statics can be either immutable or mutable just like statics outside of external blocks. An immutable static must be initialized before any Rust code is executed. It is not enough for the static to be initialized before Rust code reads from it. A mutable extern static is always `unsafe` to access, the same as a Rust mutable static, and as such can not be marked with a `safe` qualifier.
```

Simply declaring extern items, even without ever using them, can cause Undefined Behavior.
When performing cross-language compilation, attributes on one function declaration can flow to the foreign declaration elsewhere within LLVM and cause a miscompilation.
In Rust we consider all sources of Undefined Behavior to be `unsafe`, and so we must make declaring extern blocks be `unsafe`.
The up-side to this change is that in the new style it will be possible to declare an extern fn that's safe to call after the initial unsafe declaration.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Rust can utilize functions and statics from foreign code that are provided during linking, though it is `unsafe` to do so.

An `extern` block can be placed anywhere a function declaration could appear (generally at the top level of a module).

* On editions >= 2024, you *must* write all `extern` blocks as `unsafe extern`.
* On editions < 2024, you *may* write `unsafe extern`, or you can write an `extern` block without the `unsafe` keyword. Writing an `extern` block without the `unsafe` keyword is provided for compatibility only, and will generate a warning.

> * `unsafe extern` interacts with the `unsafe_code` lint, and a `deny` or `forbid` with that lint will deny or forbid the unsafe external block.

Reviewer 6: Will an `extern` (no `unsafe`, `edition < 2024`) block trip the lint? I'd argue it's as unsafe (if not more dangerous) with `unsafe` token...

Reviewer: for compatibility reasons I'd expect to warn at most, even if you use deny/forbid unsafe_code.

Author: My intent is that you get a forward compatibility warning if you write `extern` without `unsafe`.

Within an `extern` block is zero or more declarations of external functions and/or external static values.
An extern function is declared with a `;` instead of a function body (similar to a method of a trait).
An extern static value is also declared with a `;` instead of an expression (similar to an associated const of a trait).
In both cases, the actual function body or value is provided by whatever external source (which is probably not even written in Rust).

> When an `unsafe extern` block is used, all declarations within that `extern` block *should* have the `unsafe` or `safe` keywords as part of their signature.

Reviewer 6: ```suggestion
When an `unsafe extern` block is used, all declarations within that `extern` block *must* have the `unsafe` or `safe` keywords as part of their signature.
```

Since `unsafe extern` is new, we can _require_ that an annotation (`safe`|`unsafe`) is provided

> If one of the two keywords is not explicitly provided, the declaration is assumed to be `unsafe`, and also a warning is generated.

Reviewer: ```suggestion
```
`unsafe`/`safe` being required means we don't have to address when `unsafe`/`safe` is missing.

When coerced to a function pointer, a function declared in an extern block has type
```rust
extern "abi" for<'l1, ..., 'lm> fn(A1, ..., An) -> R
```
where `'l1`, ... `'lm` are its lifetime parameters, `A1`, ..., `An` are the declared types of its parameters and `R` is the declared return type.

### Statics
Statics within external blocks are declared in the same way as statics outside of external blocks, except that they do not have an expression initializing their value. If the static is unsafe to access, then the static should use the `unsafe` qualifier. If the static is safe to access (and immutable), then the static should use the `safe` qualifier (a contextual keyword). Statics that are not qualified as `unsafe` or `safe` are assumed to be `unsafe`.

Extern statics can be either immutable or mutable just like statics outside of external blocks. An immutable static must be initialized before any Rust code is executed. It is not enough for the static to be initialized before Rust code reads from it. A mutable extern static is always `unsafe` to access, the same as a Rust mutable static.

# Drawbacks
[drawbacks]: #drawbacks

> * It is very unfortunate to have to essentially reverse the status quo.

Reviewer 6: This seems to mention a drawback of a previous version? (where no qualifier in `unsafe extern {}` meant "safe")

Author: Uh, yeah, maybe, i dunno, i mean we're still swapping around how things work, maybe not "reversing" maybe some other verb.

I don't know how precisely to describe the motion happening, but also I don't wanna say that "there's no drawbacks at all!". it's definitely some motion.

Reviewer 6: Hm. How about something like this?

> * It is very unfortunate to change the syntax of `extern` blocks and require changes to all of them.
>   * Hopefully, allowing people to safely call some foreign functions will make up for the churn caused by this change.

Reviewer 4: Yeah I agree "reverse" is confusing, I don't know what is being reverse here. :+1: for something like Waffle's proposal. Basically the drawback is "churn".

Reviewer 2: Fixed.

One could ask, why not use `trusted` rather than `safe` as the contextual keyword?  E.g.:

```rust
unsafe extern {
    pub trusted fn sqrt(x: f64) -> f64; // Safe to call.
    pub unsafe fn strlen(p: *const c_char) -> usize;
}
```

The Rust language already has an accepted semantic for "safe" and "unsafe".  If we were to introduce a separated "trusted" concept, that would need to be part of a larger plan.  Such a plan does not yet exist in any concrete form, and it's not clear at this point whether any plan along these lines will succeed in gaining consensus.  Waiting to deliver value here on that possibility seems like a bad plan.

If we later decide, e.g., to replace all uses of `unsafe { .. }` with `trusted { .. }`, large amounts of code would need to be changed in that migration.  Changing from `safe fn` to `trusted fn` as part of that, as this RFC would require, doesn't seem that it would make that migration markedly more painful.

## Fire the `unsafe_code` lint for `extern` blocks also

This RFC specifies that the `unsafe_code` lint will fire for `unsafe extern` but not for `extern` blocks.  One could ask, why not fire this for `extern` blocks also?

> The problem with doing this is that it may be very noisy.  We're careful when expanding the meaning of existing lints to not create too much noise, and doing this, at least immediately, may run afoul of this.

Reviewer 8: This "reduce noise" argument contradicts with line 25 that says

> - On editions >= 2024, you *must* write all `extern` blocks as `unsafe extern`.

so you are forcing the noise *anyway* for 2024+ crates.

additionally line 26 also mentioned 

> - On editions < 2024, you *may* write `unsafe extern`, or you can write an `extern` block without the `unsafe` keyword. Writing an `extern` block without the `unsafe` keyword is provided for compatibility only, ***and will eventually generate a warning.***

meaning you can't escape from the noise by staying in 2021 either.

Reviewer 2: Thanks.  I've added clarification on this point.

One could ask, why not use `trusted` rather than `safe` as the contextual keyword?  E.g.:

```rust
unsafe extern {
    pub trusted fn sqrt(x: f64) -> f64; // Safe to call.
    pub unsafe fn strlen(p: *const c_char) -> usize;
}
```

The Rust language already has an accepted semantic for "safe" and "unsafe".  If we were to introduce a separated "trusted" concept, that would need to be part of a larger plan.  Such a plan does not yet exist in any concrete form, and it's not clear at this point whether any plan along these lines will succeed in gaining consensus.  Waiting to deliver value here on that possibility seems like a bad plan.

If we later decide, e.g., to replace all uses of `unsafe { .. }` with `trusted { .. }`, large amounts of code would need to be changed in that migration.  Changing from `safe fn` to `trusted fn` as part of that, as this RFC would require, doesn't seem that it would make that migration markedly more painful.

## Fire the `unsafe_code` lint for `extern` blocks also

This RFC specifies that the `unsafe_code` lint will fire for `unsafe extern` but not for `extern` blocks.  One could ask, why not fire this for `extern` blocks also?

> The problem with doing this is that it may be very noisy on existing editions.  We're careful when expanding the meaning of existing lints to not create too much noise for existing code on existing editions, and doing this, at least immediately, may run afoul of this.

Reviewer 8: There could be an alternative that `extern` in ≤2021 *won't* emit `unsafe_code`  (only compat lint) unless it contained at least one explicitly `safe`/`unsafe` item.

Three, not allowing items to be marked as `safe` would remove one of the key tangible *benefits* that the changes in this RFC provide to users.  This would reduce the motivation to make this change at all.

## Fix LLVM and don't prefix `extern` with `unsafe`

One could ask, why not fix LLVM such that incorrect signatures in an `extern` block would not result in undefined behavior in the resulting program unless those items were used in Rust code, and then not add `unsafe extern`?

There are three problems with this.

> One, it's not entirely clear that it's feasible to fix LLVM in this way.  Moreover, it's still a bit unclear to us whether or not this behavior is allowed by the C standard.  If it is allowed, then LLVM does not, arguably, need to be fixed at all.

Reviewer 2: Thanks.  I've added language to this effect.
